internal/nearcache/repairing_task.go

```diff
@@ -0,0 +1,568 @@
+	return h.metadataContainers[partition]
+}
+
+func (h *RepairingHandler) UpdateLastKnownStaleSequence(md *MetaDataContainer, partition int32) {
+	var lastReceived int64
+	var lastKnown int64
+	for {
+		lastReceived = md.Sequence()
+		lastKnown = md.StaleSequence()
+		if lastKnown >= lastReceived {
+			break
+		}
+		if md.CASStaleSequence(lastKnown, lastReceived) {
+			break
+		}
+	}
+	h.lg.Trace(func() string {
+		return fmt.Sprintf("stale sequences updated:[map=%s,partition=%d,lowerSequencesStaleThan=%d,lastReceivedSequence=%d]",
+			h.name, partition, md.StaleSequence(), md.Sequence())
+	})
+}
+
+// Handle handles a single invalidation.
+func (h *RepairingHandler) Handle(key serialization.Data, source, partition types.UUID, seq int64) error {
+	// port of: com.hazelcast.internal.nearcache.impl.invalidation.RepairingHandler#handle(com.hazelcast.internal.serialization.Data, java.util.UUID, java.util.UUID, long)
+	// Apply invalidation if it's not originated by local member/client.
+	// Since local Near Caches are invalidated immediately there is no need to invalidate them twice.
+	if source != partition {
+		if key == nil {
+			h.nc.Clear()
+		} else {
+			k, err := h.toNearCacheKey(key)
+			if err != nil {
+				return err
+			}
+			h.nc.Invalidate(k)
+		}
+	}
+	pid, err := h.getPartitionIDOrDefault(key)
+	if err != nil {
+		return err
+	}
+	h.CheckOrRepairUUID(pid, partition)
+	h.CheckOrRepairSequence(pid, seq, false)
+	return nil
+}
+
+// HandleBatch handles a batch of validations.
+func (h *RepairingHandler) HandleBatch(keys []serialization.Data, sources []types.UUID, partitions []types.UUID, seqs []int64) error {
+	// port of: com.hazelcast.internal.nearcache.impl.invalidation.RepairingHandler#handle(java.util.Collection<com.hazelcast.internal.serialization.Data>, java.util.Collection<java.util.UUID>, java.util.Collection<java.util.UUID>, java.util.Collection<java.lang.Long>)
+	// assumes len(keys) == len(source) -- len(partitions) == len(seqs)
+	sz := len(keys)
+	for i := 0; i < sz; i++ {
+		if err := h.Handle(keys[i], sources[i], partitions[i], seqs[i]); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (h *RepairingHandler) CheckOrRepairUUID(partition int32, new types.UUID) {
+	// this method may be called concurrently: anti-entropy, event service
+	if new.Default() {
+		panic("CheckOrRepairUUID: new UUID should not be the default UUID")
+	}
+	md := h.GetMetaDataContainer(partition)
+	for {
+		prev := md.UUID()
+		if prev == new {
+			break
+		}
+		if md.CASUUID(prev, new) {
+			md.ResetStaleSequence()
+			md.ResetStaleSequence()
+			h.lg.Trace(func() string {
+				return fmt.Sprintf("invalid UUID, lost remote partition data unexpectedly:[name=%s,partition=%d,prevUuid=%s,newUuid=%s]",
+					h.name, partition, prev, new)
+			})
+			break
+		}
+	}
+}
+
+func (h *RepairingHandler) CheckOrRepairSequence(partition int32, nextSeq int64, viaAntiEntropy bool) {
+	if nextSeq <= 0 {
+		panic("CheckOrRepairSequence <= 0")
+	}
```

Comment:
Same comment about assertion. Will java client throw in runtime?

Reply:
Replaced at: https://github.com/hazelcast/hazelcast-go-client/pull/807/commits/0d49f98e13b8fc0d4843db1366d353fb6a96a174